### ISSUE TITLE
0.8.2 -> 0.8.156 task shell tokenize command differently causing command falure

### ISSUE THREAD

**Issue by liny01-nbsa** [babashka/babashka#1310]
bb version: 0.8.2 (Good), 0.8.156 (Bad)
Platform: Windows

A command string passed to shell that works fine in 0.8.2 stopped working in 0.8.156.

```edn
  build-release-package
  {:doc "Generate a RELEASE flavor Web Deploy package, compatible with AWS Elastic Beanstalk."
   ;;  :depends [-clean-tree-test]
   :task (let [cmd (str "\"" (:msbuild-path (combo!)) "\""
                        " /p:Configuration=Release"
                        " /t:Package"
                        " /p:DeployIisAppPath=\"Default Web Site\"")]
           (shell cmd))}

```

Expected, and on 0.8.2: The *MSBuild* command line called successfully,

When same piece is run under 0.8.156, 
```
Microsoft (R) Build Engine version 17.2.1+52cd2da31 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

MSBUILD : error MSB1009: Project file does not exist.
Switch: Default Web Site
Error while executing task: build-release-package
```

After some digging, I believe the one passed down is:
```
"D:\\Program Files\\Microsoft Visual Studio\\2022\\Professional\\MSBuild\\Current\\Bin\\MSBuild.exe" /p:Configuration=Release /t:Package /p:DeployIisAppPath= Default Web Site 
```

The spaces around `Default Web Site` is not kept.

I'm able to make it work by change the `\"` to `\\\"` or `'`

**Comment by borkdude** [comment]
@liny01-nbsa I recommend using it like this:

```
  build-release-package
  {:doc "Generate a RELEASE flavor Web Deploy package, compatible with AWS Elastic Beanstalk."
   ;;  :depends [-clean-tree-test]
   :task (shell (:msbuild-path (combo!))  "/p:Configuration=Release" "/t:Package"  "/p:DeployIisAppPath=\"Default Web Site\""))}
```

So just pass the first argument separately. If the first argument refers to an existing file, then it won't be tokenized.

**Comment by liny01-nbsa** [comment]
Thanks, @borkdude , I will give that a try. Does this mean the calling form of `sh` and `shell` are mostly aligned now?

**Comment by borkdude** [comment]
Do you mean `clojure.java.shell/sh` ?

**Comment by liny01-nbsa** [comment]
Yes! I also use `clojure.java.shell/sh` in a few other place where I want the output, like:
```edn
(-> (sh "git" "rev-parse" "--short" "HEAD")
   :out
   str/trim)
```

From my previous reading of *Babashka book`, `shell` take a single long string for command and `sh` take all elements of command directly in the call.

It looks now all accepts the form that has the full command listed out.

**Comment by borkdude** [comment]
We need to document this better. This has always been the case with `shell`:

```
(shell option-map "program args")
;; or 
(shell option-map "program" "args" ...)
```
where the option map is optional.

The first string will be tokenized, unless the whole string refers to an existing file.

**Comment by liny01-nbsa** [comment]
I tried the method you suggested, it works like a charm, thanks!

**Comment by borkdude** [comment]
Cool. So we need to improve docs, which is tracked here: https://github.com/babashka/process/issues/75

**Comment by liny01-nbsa** [comment]
I noticed between 0.8.2 to 0.8.156, the shell code is moved to `process.shell`, does that also change the way it tokonize the string, especially how escaped double quote `"`or backslashes `\`are handled?

**Comment by borkdude** [comment]
Possibly. There is some `:escape` function that does things differently on Windows. You can disable that using `:escape identity` but I think it does the right thing in case you're passing JSON on the command line.

**Comment by liny01-nbsa** [comment]
Interesting, I will give that a try too.

**Comment by borkdude** [comment]
Maybe @bobisageek can also take a peek at this issue as he is familiar with related Windows issues in bb.

**Comment by bobisageek** [comment]
To make this a little nicer to experiment with, I made a slightly simpler repro:
```
; bb.edn
{:tasks {repro {:doc  ""
                :task (let [cmd (str "\"" "cmd" "\""
                                     " /c"
                                     " echo"
                                     " /p:Configuration=Release"
                                     " /t:Package"
                                     " /p:DeployIisAppPath=\"Default Web Site\"")]
                        (println cmd)
                        (shell cmd))}}}
```
... and this does indeed produce different outputs:
```shell
>bb version
babashka v0.8.156

>bb repro
"cmd" /c echo /p:Configuration=Release /t:Package /p:DeployIisAppPath="Default Web Site"
/p:Configuration=Release /t:Package /p:DeployIisAppPath= "Default Web Site"

>bb-2 version
babashka v0.8.2

>bb-2 repro
"cmd" /c echo /p:Configuration=Release /t:Package /p:DeployIisAppPath="Default Web Site"
/p:Configuration=Release /t:Package "/p:DeployIisAppPath=Default Web Site"
```
So, the quoting does come out differently. 

I'm going to see if I can dig out the cause(s) of this change.

**Comment by bobisageek** [comment]
It appears as though this behavior was changed by https://github.com/babashka/process/commit/aabbca2e7e8ccabe579f4953ca9ea1147a30ff06

The difference in behavior appears to be coming strictly from tokenize (so I don't think there's any difference between platforms). An even simpler repro (which I'm leaving as a task for the sake of not dealing with shell-specific quote escaping):

```clj
; bb.edn
{:tasks {repro (prn (babashka.process/tokenize "something=\"something else\""))}}
```
```shell
>bb repro
["something=" "something else"]

>bb-2 repro
["something=something else"]
```

I think it could _maybe_ be argued that neither of these is correct??? I imagine the desired behavior would be that:
```
(tokenize "something=\"something else\"") -> ["something=\"something else\""]
; or...
something="something else" -> something="something else"
```
and I assume that this is the desired behavior for this intermediate result/command line, e.g.:
`"...MSBuild.exe" /p... /p:DeployIisAppPath="Default Web Site"`

but I could easily be wrong and/or there might be conflicts with another use case that prompted the aforementioned change.
@borkdude or @liny01-nbsa , does this help? I don't want to propose/work on any change without making sure I've got the context right first. :)
```

**Comment by borkdude** [comment]
I think you're correct and this may be a bug in tokenize. Thanks for reproducing this! Fix welcome.

**Comment by liny01-nbsa** [comment]
> and I assume that this is the desired behavior for this intermediate result/command line, e.g.: 
> 
>`"...MSBuild.exe" /p... /p:DeployIisAppPath="Default Web Site"`
> 

Yes, @bobisageek , I think the above is the desired behavior in my use case.

**Comment by borkdude** [comment]
Didn't quite cover it yet, but I'm posting some notes here.

My hunch is that tokenize should only split if the character before `\"` was a whitespace. Maybe commit https://github.com/babashka/process/commit/aabbca2e7e8ccabe579f4953ca9ea1147a30ff06 should be reverted.
To get the last character from the stringwriter, you can use this:

```
(let [buf (-> buf (.getBuffer))
                            len (.length buf)
                            sep (when (pos? len)
                                  (.charAt buf (dec len)))]
                        sep)
```

Not sure if that's necessary.

**Comment by borkdude** [comment]
The behavior should be as printed here:

```
 $ bb -e '*command-line-args*' -- some="something else"
("some=something else")
```
